libsolidity/analysis/ImmutableValidator.cpp

```diff
@@ -206,29 +216,32 @@ void ImmutableValidator::analyseVariableReference(VariableDeclaration const& _va
 				_expression.location(),
 				"Cannot write to immutable here: Immutable variables cannot be initialized inside an if statement."
 			);
-		else if (m_initializedStateVariables.count(&_variableReference))
+		else if (m_initializedStateVariables.count(&_variableReference) || _variableReference.value())
 			m_errorReporter.typeError(
 				1574_error,
 				_expression.location(),
 				"Immutable state variable already initialized."
 			);
 		m_initializedStateVariables.emplace(&_variableReference);
 	}
-	if (read && m_inConstructionContext)
+	if (
+		read &&
+		m_inConstructionContext &&
+		(!m_initializedStateVariables.count(&_variableReference) || write)
```

Comment:
Is the `|| write` needed?

Reply:
I think it is. quite a few tests behave differently without it:

```
syntaxTests/immutable/delete_and_initialize.sol: FAIL
  Contract:
    contract C {
        uint immutable x = 3;
        constructor() {
            delete x;
        }
    }

  Expected result:
    TypeError 1574: (74-75): Immutable state variable already initialized.
    TypeError 7733: (74-75): Immutable variables cannot be read before they are in
itialized.
  Obtained result:
    TypeError 1574: (74-75): Immutable state variable already initialized.


  Contract:
    contract C {
        uint immutable x;
        constructor() {
            x++;
        }
    }

  Expected result:
    TypeError 7733: (63-64): Immutable variables cannot be read before they are in
itialized.
  Obtained result:
    Success

syntaxTests/immutable/delete.sol: Exception during test: ../libsolidity/codegen/LV
alue.cpp(187): Throw in function virtual void solidity::frontend::ImmutableItem::s
etToZero(const solidity::langutil::SourceLocation&, bool) const
Dynamic exception type: boost::wrapexcept<solidity::langutil::InternalCompilerErro
r>
std::exception::what: Attempted to set immutable variable to zero.
[solidity::util::tag_comment*] = Attempted to set immutable variable to zero.

syntaxTests/immutable/increment_decrement.sol: FAIL
  Contract:
    contract C {
      uint immutable x;
      uint immutable y;
      constructor() {
        ++x;
        --y;
      }
    }

  Expected result:
    TypeError 7733: (77-78): Immutable variables cannot be read before they are in
itialized.
    TypeError 7733: (86-87): Immutable variables cannot be read before they are in
itialized.
  Obtained result:
    Success

```

Reply:
I think the `read && write` case deserves its own error message